pkg/pillar/zedcloud/authen.go

```diff
@@ -0,0 +1,508 @@
+// Copyright (c) 2020 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// Common code to communicate to zedcloud
+
+package zedcloud
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"github.com/golang/protobuf/proto"
+	zauth "github.com/lf-edge/eve/api/go/auth"
+	zcert "github.com/lf-edge/eve/api/go/certs"
+	etpm "github.com/lf-edge/eve/pkg/pillar/evetpm"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	"github.com/satori/go.uuid"
+	log "github.com/sirupsen/logrus"
+	"io/ioutil"
+	"math/big"
+	"mime"
+	"net/http"
+	"os"
+)
+
+var onBoardCert, deviceCert *tls.Certificate
+var serverLeafCert *x509.Certificate
+var deviceCertHash, onBoardCertHash, cloudCertHash []byte
+var onBoardCertBytes []byte
+
+const (
+	hashSha256Len16 = 16 // senderCertHash size of 16
+	hashSha256Len32 = 32 // size of 32 bytes
+)
+
+// given an envelope protobuf received from controller, verify the authentication
+func verifyAuthentication(c []byte, skipVerify bool) ([]byte, types.HTTPExtraStatus, error) {
+	extraSt := types.ExtStatusNone
+	sm := &zauth.AuthContainer{}
+	err := proto.Unmarshal(c, sm)
+	if err != nil {
+		log.Infof("verifyAuthentication: can not unmarshal authen content, %v\n", err)
+		return nil, extraSt, err
+	}
+
+	data := sm.AuthPayload.GetPayload()
+	if !skipVerify { // no verify for /certs itself
+		if len(sm.GetSenderCertHash()) != hashSha256Len16 &&
```

Comment:
This code doesn't fail if the sm.Algo is unknown.
I suggest restructuring it as a switch statement on sm.Algo (and with the default returning an error).
Then each part of the swiytchj can check the length etc for the specific algo instread of these and/or conditions in this block.

Reply:
In fact it does check, but is it too hard to read to see that. 
Thus another argument to structure it as a switch statement.

Reply:
ok.